### ISSUE TITLE
Allowing derived values from override fields 

### ISSUE THREAD

**Issue by rob-p** [tweag/nickel#1682]
I wasn't quite clear if this is a bug report, or a feature request, or something else entirely, so please feel free to move/change the tag as appropriate.

We have a nickel configuration file, where many fields are derived from some top-level field. For example, something like (call this `override.ncl`):

```
$ cat override.ncl
let foo = "bar"
in
let top = {
  derived_one = {
    inner = {
      qux1 = "%{foo}/derived1"
    }
  },
  derived_two = {
    inner = {
      qux2 = "%{foo}/derived1"
    }
  }
} in top.derived_one & top.derived_two
```

this derives to (exported as json)

```
$ nickel export -f override.ncl
{
  "inner": {
    "qux1": "bar/derived1",
    "qux2": "bar/derived1"
  }
}
```

Now, we'd like to let the user override the value of `foo` from the command line when evaluating the above.  For example `--override foo \"baz\"`.  Ideally, this would force the re-evaluation of the dependent values for this term, and result in having `foo` take the value `"baz"`, **and similarly affecting the derived values of `derived_one.qux` and `derived_two.qux`**.

However, when such an override is applied, it seems to only directly affect the value of `foo`, but `derived_one.qux` and `derived_two.qux` remain as `bar/derived_1` and `bar/derived_2` respectively.  It derives to the json:

```
$ nickel export -f override.ncl -- --override foo \"qux\"
{
  "inner": {
    "qux1": "bar/derived1",
    "qux2": "bar/derived1"
  }
}
```

Any idea how to achieve the former behavior?  By the way, we are not restricted to just using the nickel CLI, and are happy to write some code to achieve this.

Thanks!
Rob

P.S.  I believe it has something to do with my current lack of understanding of how the let bindings and & interact. If I take the following:

```
let top = {
  foo = "bar",
  derived_one = {
    inner = {
      qux1 = "%{foo}/derived1"
    }
  },
  derived_two = {
    inner = {
      qux2 = "%{foo}/derived1"
    }
  }
} in top & top.derived_one & top.derived_two
```

and then run the command 

```
$ nickel export -f override.ncl -- --override foo \"qux\"
```

I get 

```
{
  "derived_one": {
    "inner": {
      "qux1": "qux/derived1"
    }
  },
  "derived_two": {
    "inner": {
      "qux2": "qux/derived1"
    }
  },
  "foo": "qux",
  "inner": {
    "qux1": "bar/derived1",
    "qux2": "bar/derived1"
  }
}
```

So you can see that in the expansion of `top` itself, the replacement works as (I) expected, but in the merged expansion of `top.derived_one & top.derived_two`, we have the original, non-overridden value of `foo`.

**Comment by yannham** [comment]
It's very interesting that you're trying to override a let-binding, because this means that this is not obvious from the documentation that you can't do so. Thanks for pointing this out, I think this might lead to stress this point somewhere in the manual.

In Nickel, let-bindings are really just normal, local, constant definitions, as in many other programming languages. It's a basic way to avoid repetition, but let-bindings aren't exposed to overriding.

In fact, I can't reproduce your first result, because here is what I get on master (with the new CLI syntax):

```console
$ nickel export test-override.ncl -- --override foo \"qux\"
error: invalid override: unknown field `foo`
 = `foo` doesn't refer to record field accessible from the root of the configuration and thus can't be the target of `--override`.
```

What version are you using? Is it the latest released 1.2.2?

What you can override are records with inter-dependencies. I think the most straightforward way to do so is thus to just have one big record:

```nickel
{
  foo
    | String
    | not_exported
    = "bar",

  derived_one.inner.qux1 = "%{foo}/derived1",
  derived_two.inner.qux2 = "%{foo}/derived1",
}
```

You now get the expected result:

```console
$ nickel export test-override.ncl -- --override foo \"qux\"
{
  "derived_one": {
    "inner": {
      "qux1": "qux/derived1"
    }
  },
  "derived_two": {
    "inner": {
      "qux2": "qux/derived1"
    }
  }
}
```

We used `not_exported` to tell Nickel that `foo` is somehow a parameter, but shouldn't appear in the final exported version. As long as you don't export and stay within the Nickel boundaries, it's just a normal record field.

If you want to split your configuration into parts and merge them, you can do it like that [^1]:

```nickel
# input.ncl
{
  foo | not_exported = "bar",
}

# derived1.ncl
{foo, derived_one.inner.qux1 = "%{foo}/derived1" }

# derived2.ncl
{foo, derived_two.inner.qux2 = "%{foo}/derived1" }
```

```console
$ nickel export input.ncl derived1.ncl derived2.ncl -- --override foo \"qux\"
{
  "derived_one": {
    "inner": {
      "qux1": "qux/derived1"
    }
  },
  "derived_two": {
    "inner": {
      "qux2": "qux/derived1"
    }
  }
}
```

In this example, when writing `derived_xx`, `foo` isn't in scope anymore. Nickel doesn't have dynamic scoping on purpose (although merging has certainly such a flavor, but it's not exactly the same thing) - to avoid using unbound names by accident, among many other things. So, a bit like a function parameter, you have to bring `foo` into scope: the simplest is to add the field to the record, without any definition nor additional metadata. This is a way of saying: I expect to be merged with `foo` in the final result. If not, Nickel will raise a missing field definition error. As long as I don't try to evaluate or export `derived_xxx`, it's perfectly fine to manipulate partially defined configurations.

[^1]: The CLI syntax is the one of current master and next 1.3.0 version. You can provide several files (previously only one through `-f`), and the semantics is to merge them. A 1.2.2 equivalent version is: `$ nickel export -- --override foo \"qux\" <<< '(import "input.ncl") & (import "derived1.ncl") & (import "derived2.ncl")'`

**Comment by rob-p** [comment]
Thanks for the detailed reply!  Yes, we eventually resolved this by having one big record, but then simply tagging the stuff that we derive other variables from, but which, themselves need not be exposed as `not_exported`.

I also agree it would be helpful to figure out how to make the behavior of `let` bindings clearer in the documentation. It's one of the few places (so far) where my understanding differed from what was actually implemented.

Thanks again for taking the time to answer in such detail!
